### PR TITLE
Add memspace "highest bandwidth"

### PR DESCRIPTION
**Pull request by kswiecicki**
### Description
This memspace contains memory targets ordered by the highest bandwidth value (targets higher in the ordering are prioritised).
There is also an alternative way to assign nodes to this memspace by setting `UMF_MEMSPACE_HIGHEST_BANDWIDTH` environment variable.

### TODO
- [x] Add tests for "highest bandwidth" memspace from `UMF_MEMSPACE_HIGHEST_BANDWIDTH`
- [x] Verify functionality on QEMU runner

### Checklist

- [x] Code compiles without errors locally
- [x] All tests pass locally
- [x] CI workflows execute properly

- [x] CI workflows, not executed per PR (e.g. Nightly), execute properly 
- [x] New tests added, especially if they will fail without my changes
- [x] Extended the README/documentation
- [x] All newly added source files have a license
- [x] All newly added source files are referenced in CMake files
- [x] Logger (with debug/info/... messages) is used

### CODE REVIEW

**Comment by kswiecicki** [comment]
To use HWLOC property query API, we need the the lib version `>= 2.3.0`. Unfortunately on the `Ubuntu 20.04` runners the latest available package is version `2.1.0` (https://packages.ubuntu.com/search?suite=all&searchon=names&keywords=hwloc). @lukaszstolarczuk do you think we could use a script to build and install HWLOC from source instead of relying on the package manager? Or alternatively install the package from newer repo eg. `jammy (22.04LTS)`?

**Comment by lukaszstolarczuk** [comment]
> To use HWLOC property query API, we need the the lib version `>= 2.3.0`. Unfortunately on the `Ubuntu 20.04` runners the latest available package is version `2.1.0` (https://packages.ubuntu.com/search?suite=all&searchon=names&keywords=hwloc). @lukaszstolarczuk do you think we could use a script to build and install HWLOC from source instead of relying on the package manager? Or alternatively install the package from newer repo eg. `jammy (22.04LTS)`?

heh, we are using Ubuntu 22.04. I think the error you have is caused because of typo - you specified to look for `>=2.30.0` 😉 
We have 2.7.0 installed: https://github.com/oneapi-src/unified-memory-framework/pull/408/checks#step:6:189

**Comment by kswiecicki** [comment]
> > To use HWLOC property query API, we need the the lib version `>= 2.3.0`. Unfortunately on the `Ubuntu 20.04` runners the latest available package is version `2.1.0` (https://packages.ubuntu.com/search?suite=all&searchon=names&keywords=hwloc). @lukaszstolarczuk do you think we could use a script to build and install HWLOC from source instead of relying on the package manager? Or alternatively install the package from newer repo eg. `jammy (22.04LTS)`?
> 
> heh, we are using Ubuntu 22.04. I think the error you have is caused because of typo - you specified to look for `>=2.30.0` 😉 We have 2.7.0 installed: https://github.com/oneapi-src/unified-memory-framework/pull/408/checks#step:6:189

My bad, I've left the version `2.30.0` as I was testing the `FindLIBHWLOC.cmake` module.
We are also using Ubuntu 20.04 runners that use the HWLOC version `2.1.0` installed via apt package manager.

**Comment by lukaszstolarczuk** [comment]
> > > To use HWLOC property query API, we need the the lib version `>= 2.3.0`. Unfortunately on the `Ubuntu 20.04` runners the latest available package is version `2.1.0` (https://packages.ubuntu.com/search?suite=all&searchon=names&keywords=hwloc). @lukaszstolarczuk do you think we could use a script to build and install HWLOC from source instead of relying on the package manager? Or alternatively install the package from newer repo eg. `jammy (22.04LTS)`?
> > 
> > 
> > heh, we are using Ubuntu 22.04. I think the error you have is caused because of typo - you specified to look for `>=2.30.0` 😉 We have 2.7.0 installed: https://github.com/oneapi-src/unified-memory-framework/pull/408/checks#step:6:189
> 
> My bad, I've left the version `2.30.0` as I was testing the `FindLIBHWLOC.cmake` module. We are also using Ubuntu 20.04 runners that use the HWLOC version `2.1.0` installed via apt package manager.

Ok, right. For the few jobs using Ubuntu 20.04 we can do what you proposed - install from source. If this will end up being more than a few lines of code, please make this a python script, e.g. `install-hwloc.py`

File: src/memory_targets/memory_target_numa.c [hunk 1]
```diff
@@ -195,12 +195,47 @@ static umf_result_t numa_get_capacity(void *memTarget, size_t *capacity) {
...
+    }
+
+    umf_result_t umf_ret = UMF_RESULT_SUCCESS;
+
+    hwloc_const_cpuset_t cCpuset = hwloc_topology_get_allowed_cpuset(topology);
+    hwloc_cpuset_t cpuset = hwloc_bitmap_dup(cCpuset);
+    struct hwloc_location initiator = {.location.cpuset = cpuset,
+                                       .type = HWLOC_LOCATION_TYPE_CPUSET};
+    hwloc_uint64_t value = 0;
+    int ret = hwloc_memattr_get_value(topology, HWLOC_MEMATTR_ID_BANDWIDTH,
```
Comment (igchor) [inline src/memory_targets/memory_target_numa.c:218]: What will this function return if there are multiple nodes in the cpuset? Shouldn't `numa_get_bandwidth` accept an initiator node? We could then look at bandwidth between every pair of nodes in init code. I think we do that in memkind already:

https://github.com/memkind/memkind/blob/c425635feb0367c43da8fa0deadb50a4bc127988/src/memkind_mem_attributes.c#L324

Comment (kswiecicki) [reply src/memory_targets/memory_target_numa.c:218]: `numa_get_bandwidth` accepts a `memTarget` which represents a node, in this case the initiator is the cpuset on the platform. Do you mean that we should get the `cpuset` information from the HWLOC numa node obj structure?

Comment (igchor) [reply src/memory_targets/memory_target_numa.c:218]: I was suggesting that perhaps we should accept both `initiator` and `target` nodes as params in `numa_get_bandwidth` because we are interested in bandwidth between every pair of nodes I think.

**Comment by kswiecicki** [comment]
The original implementation of the `highest bandwidth` memspace was supposed to contain NUMA nodes (represented as memory targets in UMF) sorted by either the average or maximum (it was not yet decided) bandwidth value. This value was obtained by considering each NUMA node as the initiator for the bandwidth query and selecting all NUMA nodes as targets consecutively.
Since control over allocation placement is restricted to the nodemask provided to the membind function in the OS provider, we were unable to prioritize NUMA nodes with better bandwidth values. For now, we've settled on the `highest bandwidth` memspace implementation that contains an aggregated list of NUMA nodes identified as the best targets after selecting each NUMA node as the initiator.

This comment serves as documentation for the development of the memspace API.
@igchor would you like to add something?

File: .github/scripts/install_hwloc.sh [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+# Copyright (C) 2024 Intel Corporation
+# Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+# install_hwloc.sh - Script for building and installing HWLOC library from source code
+
+set -e
+
+git clone -b hwloc-2.10.0 https://github.com/open-mpi/hwloc.git
```
Comment (bratpiorka) [inline .github/scripts/install_hwloc.sh:10]: you wrote in reqs that we need 2.3 but you install 2.10?

File: src/memspaces/memspace_highest_bandwidth.c [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+static void umfMemspaceHighestBandwidthInit(void) {
+    umf_result_t ret =
+        umfMemspaceHighestBandwidthCreate(&UMF_MEMSPACE_HIGHEST_BANDWIDTH);
+    if (ret != UMF_RESULT_SUCCESS) {
+        assert(ret == UMF_RESULT_ERROR_NOT_SUPPORTED);
+    }
+
+    // TODO: Setup appropriate cleanup when HBW memspace becomes available
+    // on Windows. HBW memspace depends on OS provider, which currently
```
Comment (bratpiorka) [inline src/memspaces/memspace_highest_bandwidth.c:139]: OS provider is supported on windows

File: .github/scripts/install_hwloc.sh (hunk 2 above)
Comment (kswiecicki) [reply .github/scripts/install_hwloc.sh:10]: Well, I've tested it with 2.3.0, but 2.10.0 is the latest version available.

File: cmake/FindLIBHWLOC.cmake [hunk 4]
```diff
@@ -14,6 +14,23 @@ find_file(LIBHWLOC_HEADER NAMES hwloc.h)
...
+    else()
+        message(
+            WARNING
+                "LIBHWLOC_INCLUDE_DIR found, but header with version info is missing"
+        )
+    endif()
+endif()
+
 if(WINDOWS)
     find_file(LIBHWLOC_DLL NAMES "bin/hwloc-15.dll" "bin/libhwloc-15.dll")
```
Comment (ldorau) [inline cmake/FindLIBHWLOC.cmake:35]: Maybe it would be more safe to replace it with `NAMES "bin/hwloc-*.dll" "bin/libhwloc-*.dll"` ?

**Review by ldorau — CHANGES_REQUESTED** [review]

File: cmake/FindLIBHWLOC.cmake (hunk 4 above)
Comment (ldorau) [reply cmake/FindLIBHWLOC.cmake:35]: See: #426

File: cmake/FindLIBHWLOC.cmake [hunk 5]
```diff
@@ -14,6 +14,23 @@ find_file(LIBHWLOC_HEADER NAMES hwloc.h)
 get_filename_component(LIBHWLOC_INCLUDE_DIR ${LIBHWLOC_HEADER} DIRECTORY)
 set(LIBHWLOC_INCLUDE_DIRS ${LIBHWLOC_INCLUDE_DIR})
 
+if(LIBHWLOC_INCLUDE_DIR)
```
Comment (ldorau) [inline cmake/FindLIBHWLOC.cmake:17]: Please split this PR and extract all HWLOC stuff into a separate PR.

**Review by ldorau — CHANGES_REQUESTED** [review]

File: cmake/FindLIBHWLOC.cmake (hunk 5 above)
Comment (kswiecicki) [reply cmake/FindLIBHWLOC.cmake:17]: Done, https://github.com/oneapi-src/unified-memory-framework/pull/432.

File: cmake/FindLIBHWLOC.cmake (hunk 4 above)
Comment (kswiecicki) [reply cmake/FindLIBHWLOC.cmake:35]: It seems that the `find_file` doesn't accept regex nor wildcards. I've considered using `file(GLOB/GLOB_RECURSE ...)`, but I would need to rely on some root dir. Maybe using parent directory of `${LIBHWLOC_LIB_DIR}` could work, but I'm not sure if its consistent solution.

Comment (ldorau) [reply cmake/FindLIBHWLOC.cmake:35]: Thanks

**Comment by igchor** [comment]
@kswiecicki could you add a test on qemu that uses the highest bandwidth memspace and verifies that in case of equal bandwidths between nodes, local node is preferred?

The test should spawn multiple threads and pin every thread to a different NUMA node and the allocate memory from each thread. The allocated memory should come from the local node.

File: README.md [hunk 6]
```diff
@@ -213,6 +213,11 @@ using umfMemspaceHostAllGet.
 Memspace backed by all available NUMA nodes discovered on the platform sorted by capacity.
 Can be retrieved using umfMemspaceHighestCapacityGet.
 
+#### Highest bandwidth memspace
+
+Memspace backed by an aggregated list of NUMA nodes identified as best targets after selecting each available NUMA node as the initiator.
```
Comment (bratpiorka) [inline README.md:218]: "best targets" -> "highest bandwidth targets" or something more precise

**Review by ldorau — CHANGES_REQUESTED** [review]

File: src/memory_target.c [hunk 7]
```diff
@@ -83,3 +83,14 @@ umf_result_t umfMemoryTargetGetCapacity(umf_memory_target_handle_t memoryTarget,
     assert(capacity);
     return memoryTarget->ops->get_capacity(memoryTarget->priv, capacity);
 }
+
+umf_result_t
+umfMemoryTargetGetBandwidth(umf_memory_target_handle_t srcMemoryTarget,
+                            umf_memory_target_handle_t dstMemoryTarget,
+                            size_t *bandwidth) {
+    assert(srcMemoryTarget);
```
Comment (bratpiorka) [inline src/memory_target.c:91]: I see that this is not a common practice in this file but we should check for arguments using ifs and return UMF_RESULT_ERROR_INVALID_ARGUMENT when needed

File: src/memory_targets/memory_target_numa.c [hunk 8]
```diff
@@ -143,12 +144,57 @@ static umf_result_t numa_get_capacity(void *memTarget, size_t *capacity) {
     return UMF_RESULT_SUCCESS;
 }
 
+static umf_result_t numa_get_bandwidth(void *srcMemoryTarget,
+                                       void *dstMemoryTarget,
+                                       size_t *bandwidth) {
+    hwloc_topology_t topology = umfGetTopology();
```
Comment (bratpiorka) [inline src/memory_targets/memory_target_numa.c:150]: check if args != NULL

File: src/memspace.c [hunk 9]
```diff
@@ -208,3 +208,89 @@ umfMemspaceSortDesc(umf_memspace_handle_t hMemspace,
...
+    umf_memspace_handle_t newMemspace =
+        umf_ba_global_alloc(sizeof(struct umf_memspace_t));
+    if (!newMemspace) {
+        ret = UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+        goto err_free_best_targets;
+    }
+
+    newMemspace->size = numUniqueBestNodes;
+    newMemspace->nodes = umf_ba_global_alloc(
+        sizeof(umf_memory_target_handle_t) * newMemspace->size);
```
[the remark this answers is not in the document]
Comment (bratpiorka) [reply src/memspace.c:265]: why? I think it is more readable (and easier to check for potential errors) to pass type to sizeof()

File: src/memspaces/memspace_highest_bandwidth.c [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+
+#define MAX_NODES 512
+
+static bool is_number_str(const char *token) {
+    if (*token == '\0') {
+        return false;
+    }
+
+    while (*token != '\0') {
+        if (*token < '0' || *token > '9') {
```
Comment (bratpiorka) [inline src/memspaces/memspace_highest_bandwidth.c:31]: you could use isdigit(*token)

File: src/memspace.c [hunk 11]
```diff
@@ -208,3 +208,89 @@ umfMemspaceSortDesc(umf_memspace_handle_t hMemspace,
 
     return UMF_RESULT_SUCCESS;
 }
+
+enum umf_result_t umfMemspaceFilter(
```
[the remark this answers is not in the document]
Comment (kswiecicki) [reply src/memspace.c]: Done.

File: src/memspace.c [hunk 12]
```diff
@@ -208,3 +208,89 @@ umfMemspaceSortDesc(umf_memspace_handle_t hMemspace,
...
+                              umf_memory_target_handle_t *nodes,
+                              size_t numNodes,
+                              umf_memory_target_handle_t *target),
+    umf_memspace_handle_t *filteredMemspace) {
+    if (!hMemspace || !getTarget) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    umf_memory_target_handle_t *uniqueBestNodes = umf_ba_global_alloc(
+        hMemspace->size * sizeof(umf_memory_target_handle_t *));
```
[the remark this answers is not in the document]
Comment (kswiecicki) [reply src/memspace.c]: Done.

File: src/memspace.c [hunk 13]
```diff
@@ -208,3 +208,89 @@ umfMemspaceSortDesc(umf_memspace_handle_t hMemspace,
...
+
+        // if the target is not present, add it to the best nodes
+        if (bestTargetIdx == numUniqueBestNodes) {
+            uniqueBestNodes[numUniqueBestNodes++] = target;
+        }
+    }
+
+    // copy the unique best nodes into a new memspace
+    umf_memspace_handle_t newMemspace =
+        umf_ba_global_alloc(sizeof(struct umf_memspace_t));
```
[the remark this answers is not in the document]
Comment (kswiecicki) [reply src/memspace.c]: Done.

File: test/memspaces/memspace_highest_bandwidth.cpp [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+
+    umfMemoryProviderDestroy(hProvider);
+}
+
+TEST_F(memspaceHighestBandwidthProviderTest, allocFree) {
+    void *ptr = nullptr;
+    size_t size = SIZE_4K;
+    size_t alignment = 0;
+
+    enum umf_result_t ret =
```
[the remark this answers is not in the document]
Comment (kswiecicki) [reply test/memspaces/memspace_highest_bandwidth.cpp]: Done.

File: test/memspaces/memspace_numa.cpp [hunk 15]
```diff
@@ -9,6 +9,47 @@
 
 #include <umf/providers/provider_os_memory.h>
 
+struct memspaceNumaTest : ::numaNodesTest {
+    void SetUp() override {
+        ::numaNodesTest::SetUp();
+
+        enum umf_result_t ret = umfMemspaceCreateFromNumaArray(
```
[the remark this answers is not in the document]
Comment (kswiecicki) [reply test/memspaces/memspace_numa.cpp]: Done.

File: src/memspace.c [hunk 16]
```diff
@@ -208,3 +208,89 @@ umfMemspaceSortDesc(umf_memspace_handle_t hMemspace,
 
     return UMF_RESULT_SUCCESS;
 }
+
+enum umf_result_t umfMemspaceFilter(
+    umf_memspace_handle_t hMemspace,
+    umf_result_t (*getTarget)(umf_memory_target_handle_t initiator,
```
[the remark this answers is not in the document]
Comment (kswiecicki) [reply src/memspace.c]: Done.

File: src/memspace.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (kswiecicki) [reply src/memspace.c]: Done.

File: src/memspace_internal.h [hunk 17]
```diff
@@ -39,6 +39,18 @@ typedef umf_result_t (*umfGetPropertyFn)(umf_memory_target_handle_t,
...
 
+typedef umf_result_t (*umfGetTargetFn)(umf_memory_target_handle_t initiator,
+                                       umf_memory_target_handle_t *nodes,
+                                       size_t numNodes,
+                                       umf_memory_target_handle_t *target);
+
+///
+/// \brief Filters the targets using getTarget() to create a new memspace
+///
+enum umf_result_t umfMemspaceFilter(umf_memspace_handle_t hMemspace,
```
[the remark this answers is not in the document]
Comment (kswiecicki) [reply src/memspace_internal.h]: Done.

File: scripts/qemu/run-build.sh [hunk 18]
```diff
@@ -34,7 +34,7 @@ cmake .. \
 
 make -j $(nproc)
 
-ctest --output-on-failure
+ctest --output-on-failure --verbose
```
[the remark this answers is not in the document]
Comment (kswiecicki) [reply scripts/qemu/run-build.sh:37]: Done.

File: src/memory_target.c (hunk 7 above)
Comment (kswiecicki) [reply src/memory_target.c:91]: Done.

File: src/memory_targets/memory_target_numa.c (hunk 8 above)
Comment (kswiecicki) [reply src/memory_targets/memory_target_numa.c:150]: Done.

File: src/memspaces/memspace_highest_bandwidth.c (hunk 10 above)
Comment (kswiecicki) [reply src/memspaces/memspace_highest_bandwidth.c:32]: Done.

**Comment by kswiecicki** [comment]
> @kswiecicki could you add a test on qemu that uses the highest bandwidth memspace and verifies that in case of equal bandwidths between nodes, local node is preferred?
> 
> The test should spawn multiple threads and pin every thread to a different NUMA node and the allocate memory from each thread. The allocated memory should come from the local node.

Done. I've added the test case, but there's no QEMU configuration that satisfies the equal bandwidth requirement. @KFilipek, would I need to create a separate job in `qemu.yml` workflow to run a new QEMU config?

**Review by igchor — APPROVED** [review]

**Review by bratpiorka — APPROVED** [review]

**Review by ldorau — APPROVED** [review]

**Review by KFilipek — APPROVED** [review]

File: src/memspaces/memspace_highest_bandwidth.c [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+umfMemspaceHighestBandwidthCreate(umf_memspace_handle_t *hMemspace) {
+    if (!hMemspace) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    // Check env var 'UMF_MEMSPACE_HIGHEST_BANDWIDTH' for the nodes assigned
+    // by the user.
+    unsigned nodeIds[MAX_NODES];
+    size_t nNodes = 0;
+    umf_result_t ret = getenv_to_nodes_array("UMF_MEMSPACE_HIGHEST_BANDWIDTH",
```
Comment (lplewa) [inline src/memspaces/memspace_highest_bandwidth.c:110]: Undocumented feature, and untested. 

BTW why do we need this feature?

**Review by lplewa — CHANGES_REQUESTED** [review]

File: src/memspaces/memspace_highest_bandwidth.c (hunk 19 above)
Comment (kswiecicki) [reply src/memspaces/memspace_highest_bandwidth.c:110]: It was meant to be give the user a flexibility of setting the memspace himself when there's no HMAT support on his platform. I can add it in separate PR if necessary.

File: src/memory_target.c [hunk 20]
```diff
@@ -79,7 +79,26 @@ umf_result_t umfMemoryTargetClone(umf_memory_target_handle_t memoryTarget,
 
 umf_result_t umfMemoryTargetGetCapacity(umf_memory_target_handle_t memoryTarget,
                                         size_t *capacity) {
+    if (!memoryTarget || !capacity) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
     assert(memoryTarget);
```
[the remark this answers is not in the document]
Comment (kswiecicki) [reply src/memory_target.c]: Done.

File: src/memspaces/memspace_highest_bandwidth.c (hunk 19 above)
Comment (kswiecicki) [reply src/memspaces/memspace_highest_bandwidth.c]: Done.

**Review by lukaszstolarczuk — COMMENTED** [review]
👍

**Review by lplewa — CHANGES_REQUESTED** [review]
Do those functions works on windows - you did not updated .def file.

**Comment by kswiecicki** [comment]
> Do those functions works on windows - you did not updated .def file.

Memspace functionality is currently built only on Linux platforms.

**Review by lplewa — APPROVED** [review]
